server/src/main/java/org/elasticsearch/index/search/MatchQuery.java

```diff
@@ -253,22 +253,17 @@ public Query parse(Type type, String fieldName, Object value) throws IOException
         }
         final String field = fieldType.name();
 
+        Analyzer analyzer = getAnalyzer(fieldType, type == Type.PHRASE);
+        assert analyzer != null;
+
         /*
-         * If the user forced an analyzer we really don't care if they are
-         * searching a type that wants term queries to be used with query string
-         * because the QueryBuilder will take care of it. If they haven't forced
-         * an analyzer then types like NumberFieldType that want terms with
-         * query string will blow up because their analyzer isn't capable of
-         * passing through QueryBuilder.
+         * For untokenized fields that use a keyword analyzer, we know that further
+         * analysis isn't needed and can immediately return a term query.
          */
-        boolean noForcedAnalyzer = this.analyzer == null;
-        if (fieldType.tokenized() == false && noForcedAnalyzer &&
-                fieldType instanceof KeywordFieldMapper.KeywordFieldType == false) {
+        if (fieldType.tokenized() == false && analyzer == Lucene.KEYWORD_ANALYZER) {
```

Comment:
I think we can remove `fieldType.tokenized() == false` and just check if the analyzer is a `keyword` analyzer.

Reply:
That makes sense, will update.